Source/Core/Core/Core.cpp

```diff
@@ -167,6 +167,9 @@ void FrameUpdateOnCPUThread()
 
 void OnFrameEnd()
 {
+#ifdef USE_RETRO_ACHIEVEMENTS
+  AchievementManager::GetInstance()->DoFrame();
+#endif  // USE_RETRO_ACHIEVEMENTS
```

Comment:
@phire created a [HookableEvent](https://github.com/dolphin-emu/dolphin/blob/af52b5a2d90bb1a8b2d3f88664cb54a9198b3a88/Source/Core/Common/HookableEvent.h) system that might be usable here (there's an [after frame event](https://github.com/dolphin-emu/dolphin/blob/19bf13d3bb2c02f00c1fb01da6f9c6e381a087e4/Source/Core/VideoCommon/VideoEvents.h#L15-L19)). I'm not sure if that's fully equivalent to this or not though.